Cypher engine and querying app

l-hoang opened the pull request:
Few years worth of work/commits in this PR re: the Cypher querying engine.

The main differences from the current master branch files aside from completely new files are the following:

- LC_CSR(-CSC)_Graph have a few degree getting functions, and CSR has a direct load-into-memory function as well.
- Pangolin libs have a bunch of ifdefs added by Roshan to facilitate turning on/off certain features in the QueryGraph.

New things include:
- libquery, where the Cypher compiler (parses cypher and changes to a format Galois engine understands) and querying engine reside.
- New submodule for a cypher parser used by the cypher compiler
- querying subdirectory in lonestar which currently contains an LDBC reader and querytest with querytest being the main addition in this PR (allows running queries on Katana Form graphs)
- changes to the result checker script to work with strings (used in querytest tests)

Note that the code will eventually need a modernization of sorts as it was written C-style to have it work with the use-case it was originally intended for. The namespacing and file layout could also use an improvement eventually. The point of this PR is to get this code out for people to use/look at ASAP.

roshandathathri commented on CMakeLists.txt:
```diff
@@ -22,6 +22,7 @@ endif()
 
 ###### Options (alternatively pass as options to cmake -DName=Value) ######
 ###### Distributed-heterogeneous features ######
+set(GALOIS_ENABLE_CYPHER OFF CACHE BOOL "Enable Cypher/query graph library support")
```
Don't think this is required but this is fine for now because enabling this changes Pangolin (library and apps).

l-hoang replied:
yes, i think pangolin apps break if this is turn on as you mentioned in the issue you just opened

roshandathathri commented on libgalois/include/galois/graphs/LC_CSR_CSC_Graph.h:
```diff
@@ -314,6 +314,15 @@ class LC_CSR_CSC_Graph
+   * @param mflag how safe the acquire should be
+   * @returns In-degree of node N
+   */
+  auto in_degree(GraphNode N) const {
```
Please remove this function. getInDegree() already exists (I added it recently to master).

roshandathathri commented on libgalois/include/galois/graphs/LC_CSR_Graph.h:
```diff
@@ -459,6 +459,15 @@ class LC_CSR_Graph :
+   * @param mflag how safe the acquire should be
+   * @returns Degree of node N
+   */
+  auto degree(GraphNode N) const {
```
Please remove this function. getDegree() already exists (I added it recently to master).

roshandathathri commented on libgalois/include/galois/graphs/LC_CSR_Graph.h:
```diff
@@ -796,10 +805,58 @@ class LC_CSR_Graph :
   const EdgeIndData& getEdgePrefixSum() const { return edgeIndData; }
 
+  //! Read topology of graph directly into allocated arrays (no edge data)
+  void readGraphTopology(int fd, size_t numNodes, size_t numEdges) {
```
Why is this needed? Please drop it if it is not used.

roshandathathri commented on libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h:
```diff
@@ -0,0 +1,510 @@
+   * @param mflag how safe the acquire should be
+   * @returns Degree of node N
+   */
+  auto degree(GraphNode N) const {
```
Please drop this function

roshandathathri commented on libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h:
```diff
@@ -0,0 +1,510 @@
+   * @param mflag how safe the acquire should be
+   * @returns Degree of node N
+   */
+  auto degree(GraphNode N, const EdgeTy& data) const {
```
degree -> getDegree

roshandathathri commented on libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h:
```diff
@@ -0,0 +1,510 @@
+   * @param mflag how safe the acquire should be
+   * @returns In-degree of node N
+   */
+  auto in_degree(GraphNode N, const EdgeTy& data) const {
```
in_degree -> getInDegree

roshandathathri commented on libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h:
```diff
@@ -0,0 +1,510 @@
+   * @param mflag how safe the acquire should be
+   * @returns In-degree of node N
+   */
+  auto in_degree(GraphNode N) const {
```
Please drop this function

l-hoang commented:
@ddn0 Should all 300+ commits be quashed into one? I think there may be some worth in revisiting older commits from this line of work as there are some older features in the history that we may want to bring back at some point (we've tried some interesting things over the years that aren't currently supported in this current state), but having a single neat commit is also a huge plus. I don't really have any *strong* opinion either way anyways, but I lean towards keeping the history.

What do you think @roshandathathri ?

roshandathathri commented:
> Should all 300+ commits be quashed into one?

I'm opposed to it. I want the revision history.

roshandathathri commented on libgalois/include/galois/graphs/LC_CSR_CSC_Graph.h:
```diff
@@ -430,6 +430,24 @@ class LC_CSR_CSC_Graph
+        galois::iterate(this->begin(), this->end()),
+        [&](unsigned v) {
+          savedInDegrees[v] =
+              std::distance(this->in_edge_begin(v), this->in_edge_end(v));
```
Shouldn't this be calling getInDegree()? 

Calling non-raw begin and end prevents the function from being const.

roshandathathri commented on libgalois/include/galois/graphs/LC_CSR_CSC_Graph.h:
```diff
@@ -430,6 +430,24 @@ class LC_CSR_CSC_Graph
+   * Returns in-degrees in a vector; useful if in-degrees need to be accessed
+   * quickly (1 memory access instead of 2 from subtracting begin and end).
+   */
+  auto countInDegrees() {
```
{ -> const {

ddn0 commented:
> Should all 300+ commits be quashed into one?... I don't really have any strong opinion either way anyways, but I lean towards keeping the history.

Without a culture of common code review and continuous testing, the value of any commit history is low for the people who weren't intimately involved with the development process. So, I'm not in a good position to value the history as you have it.

What I do care about is making sure our current state has well tested code paths, clean interfaces, etc. If there is any obvious clean up you can do now, I'd suggest you do it now so those who work on this code after you have the benefit your wisdom rather than the history of your actions.

roshandathathri commented:
> those who work on this code after you have the benefit your wisdom rather than the history of your actions

Poetic. Wow!

l-hoang commented:
FYI: I have another commit in the works that is meant to remove some dead code and does some streamlining, so @remaining reviewers you may want to wait (will be ready sometime tomorrow afternoon at the latest I hope).

l-hoang commented:
Thanks everyone; will merge after this cleanup commit I'm about to push in + after CI passes for those. The only person who really needs to check the cleanup commit will be @roshandathathri (more details on what the commit has will be provided once I push).

ddn0 commented on lonestar/CMakeLists.txt:
```diff
@@ -215,6 +215,10 @@ if(GALOIS_ENABLE_DIST)
   endfunction()
 endif()
 
+if(GALOIS_ENABLE_CYPHER)
```
Can we just enable cypher by default?

(You know me and my quest to reduce build configurations :D )

roshandathathri replied:
I agree but we cannot do this as of now because enabling cypher changes/breaks Pangolin and its apps. I'll try to fix that soon so that we can enable it by default but it won't be in this PR.

l-hoang replied:
There was a weird interdependency between this code and Pangolin in which enabling cypher broke things. I think I may have fixed it some time ago; I'll get back to you in a bit.

If everything builds fine with CYPHER enabled then yes, I have no problem with this; reducing build configs is good.

l-hoang replied:
@roshandathathri This *might* be fixed now, will check in a bit.

roshandathathri replied:
It's not just about building. Pangolin should run correctly without performance degradation. I would suggest not to attempt that in this PR.

l-hoang replied:
Sure thing.

Just checked and it *does* build and run successfully with ctest (not sure if ctest for shared mem apps check for correctness) for what it's worth. We can check performance later.

l-hoang replied:
Note I'm making it build by default now because Pangolin has been completely decoupled from the querying library (as it should be).

roshandathathri replied:
> not sure if ctest for shared mem apps check for correctness

No, it doesn't. Only distributed analytics apps verify correctness.

> Note I'm making it build by default now because Pangolin has been completely decoupled from the querying library (as it should be).

Great, thanks!

ddn0 commented on lonestar/querying/cpu/querytest/querytest.cpp:
```diff
@@ -0,0 +1,191 @@
+#include "galois/Galois.h"
```
Should I file a ticket to integrate some query tests into the CI?

l-hoang replied:
Yes; they're actually already there. I'll just need to upload the inputs/outputs to the s3 bucket and fix a few assumptions probably.  That can be taken care of separately from this PR as well.

roshandathathri replied:
Yes, I think we can do that once cypher is enabled by default (see my comment above).

l-hoang commented:
Set of commits pushed.

- removed dead/unused code/functions
- added more documentation
- **Cypher enabled by default**
- removed support for older way of loading graphs (maintaining support would be a pain in the future now that we're moving to KatanaForm)

Note that Cypher is enabled by default: I removed all coupling of Pangolin and libquery as the headers we're actually using from Pangolin don't touch the files that had the #ifdefs in question. libpangolin is pretty much in line with the master branch now.

roshandathathri commented on libquery/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+target_include_directories(galois_query PUBLIC ${CYPHER_PARSER_INCLUDE})
+
+if (CMAKE_HAVE_PTHREAD_H)
+   add_definitions(-DGALOIS_HAVE_PTHREAD)
```
Why is this required?

l-hoang replied:
I have no clue actually. I think it may have been a relic for when it was interfacing with the Python code and being built separately from libgalois.

roshandathathri commented on libquery/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+
+find_package(NUMA)
+if(NUMA_FOUND)
+  add_definitions(-DGALOIS_USE_NUMA)
```
Shouldn't this be done in the common cmake? Why is this in libquery?

l-hoang replied:
See reply for PTHREAD.

roshandathathri commented on lonestar/querying/cpu/querytest/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+add_executable(querytest querytest.cpp)
+target_link_libraries(querytest galois_query)
+# TODO shouldn't need to use this define; remove it in the future
+target_compile_definitions(querytest PRIVATE GALOIS_USE_QUERY_GRAPH_TYPE=1)
```
No longer used. Drop it.

roshandathathri commented on libquery/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+)
+
+# TODO avoid this
+target_compile_definitions(galois_query PRIVATE GALOIS_USE_QUERY_GRAPH_TYPE=1)
```
No longer used. Drop this.

l-hoang commented:
So the problem with making cypher default at the moment is that the CI workflow isn't applying the patch required for the cypher parser submodule to compile. I'll check the workflow and see how to do it.

l-hoang commented:
Merging soon; feel free to make a last check if necessary.

The workflow things I had to add seem quite wordy/bad, so we'll have to fix that in a later PR.
